src/extra/gfx.cpp

```diff
@@ -0,0 +1,343 @@
+#include <magic_enum.hpp>
+
+using namespace chainblocks;
+
+namespace gfx {
+using chainblocks::Mat4;
+
+struct BuiltinFeatureBlock {
+  enum class Id {
+    Transform,
+    BaseColor,
+    VertexColorFromNormal,
+  };
+
+  static constexpr uint32_t IdTypeId = 'feid';
+  static inline chainblocks::Type IdType = chainblocks::Type::Enum(VendorId, IdTypeId);
+  static inline chainblocks::EnumInfo<Id> IdEnumInfo{"BuiltinFeatureId", VendorId, IdTypeId};
+
+  static CBTypesInfo inputTypes() { return CoreInfo::AnyType; }
+  static CBTypesInfo outputTypes() { return FeatureType; }
+
+  static inline Parameters params{{"Id", CBCCSTR("Builtin feature id."), {IdType}}};
+  static CBParametersInfo parameters() { return params; }
+
+  Id _id{};
+  FeaturePtr *_feature{};
+
+  void setParam(int index, const CBVar &value) {
+    switch (index) {
+    case 0:
+      _id = Id(value.payload.enumValue);
+      break;
+    }
+  }
+
+  CBVar getParam(int index) {
+    switch (index) {
+    case 0:
+      return Var::Enum(_id, VendorId, IdTypeId);
+    default:
+      return Var::Empty;
+    }
+  }
+
+  void cleanup() {
+    if (_feature) {
+      FeatureObjectVar.Release(_feature);
+      _feature = nullptr;
+    }
+  }
+
+  void warmup(CBContext *context) {
+    _feature = FeatureObjectVar.New();
+    switch (_id) {
+    case Id::Transform:
+      *_feature = features::Transform::create();
+      break;
+    case Id::BaseColor:
+      *_feature = features::BaseColor::create();
+      break;
+    case Id::VertexColorFromNormal:
+      *_feature = features::DebugColor::create("normal", ProgrammableGraphicsStage::Vertex);
+      break;
+    }
+  }
+
+  CBVar activate(CBContext *context, const CBVar &input) { return FeatureObjectVar.Get(_feature); }
+};
+
+struct DrawablePassBlock {
+  static CBTypesInfo inputTypes() { return CoreInfo::AnyType; }
+  static CBTypesInfo outputTypes() { return PipelineStepType; }
+
+  static inline Type PipelineStepSeqType = Type::SeqOf(PipelineStepType);
+  static inline Parameters params{
+      {"Features", CBCCSTR("Features to use."), {Type::VariableOf(PipelineStepSeqType), PipelineStepSeqType}},
+  };
+  static CBParametersInfo parameters() { return params; }
+
+  PipelineStepPtr *_step{};
+  ParamVar _features{};
+
+  void setParam(int index, const CBVar &value) {
+    switch (index) {
+    case 0:
+      _features = value;
+      break;
+    }
+  }
+
+  CBVar getParam(int index) {
+    switch (index) {
+    case 0:
+      return _features;
+    default:
+      return Var::Empty;
+    }
+  }
+
+  void cleanup() {
+    if (_step) {
+      PipelineStepObjectVar.Release(_step);
+      _step = nullptr;
+    }
+    _features.cleanup();
+  }
+
+  void warmup(CBContext *context) {
+    _step = PipelineStepObjectVar.New();
+    _features.warmup(context);
+  }
+
+  std::vector<FeaturePtr> collectFeatures(const CBVar &input) {
+    std::vector<FeaturePtr> result;
+    std::vector<Var> featureVars(Var(_features.get()));

```

Comment:
same, cache

Reply:
I added a `Var::intoVec(std::vector<T>&)` for this one.